File: Patterns/DiamondPattern.cpp
/*
Approach:
1. We use two nested loops to iterate through each row and each column of the diamond pattern.
2. The first loop is used to iterate through the rows of the diamond. It starts from 0 and goes up to n-1 rows for the upper half of the diamond.
3. Inside the first loop, we use a nested loop to print spaces before the pattern of each row. The number of spaces decreases by 1 as we move down the rows.
4. Again inside the first loop, we use another nested loop to print the pattern of asterisks (*) for each row. The number of asterisks increases by 2 as we move down the rows.
5. After printing each row, we move to the next line using `cout << endl;`.
6. Now, we need to print the lower half of the diamond. For this, we use another loop that starts from n-2 (as we have already printed the top-most row in the first loop) and moves down to 0 rows.
7. Inside this loop, we print the spaces and pattern the same way as in the first loop.
8. Finally, we have the main function where we take input from the user for the number of rows and call the `printDiamondPattern` function.

Time Complexity: O(n^2) - as we use nested loops to iterate through each row and each column of the diamond pattern.

Space Complexity: O(1) - as we only use a constant amount of additional space for variables.

Sample Input:
Enter the number of rows: 5

Sample Output:
    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *
*/


#include <iostream>
using namespace std;

void printDiamondPattern(int n) {
    // Print the upper half of the diamond
    for (int i = 0; i < n; i++) {
        // Print spaces before the pattern for each row
        for (int j = 0; j < n - i - 1; j++) {
            cout << " ";
        }
        
        // Print the pattern for each row
        for (int j = 0; j < 2 * i + 1; j++) {
            cout << "*";
        }
        
        cout << endl; // Move to the next line
    }
    
    // Print the lower half of the diamond
    for (int i = n-2; i >= 0; i--) {
        // Print spaces before the pattern for each row
        for (int j = 0; j < n - i - 1; j++) {
            cout << " ";
        }
        
        // Print the pattern for each row
        for (int j = 0; j < 2 * i + 1; j++) {
            cout << "*";
        }
        
        cout << endl; // Move to the next line
    }
}

int main() {
    int n;
    cout << "Enter the number of rows: ";
    cin >> n;
    
    printDiamondPattern(n);

    return 0;
}

File: Patterns/diamond_pattern.go
/*
In this code, the user enters a number, and the program will print a diamond pattern with that number as the widest point.

Explanation of the approach:
- We define two helper functions: `printSpaces` and `printStars`. These functions are used to print the desired spaces and stars, respectively.
- The `printDiamondPattern` function takes an integer `n` as input. It prints the upper half and lower half of the diamond pattern using a combination of spaces and stars.
- In the upper half, the number of spaces in each row decreases by 1, while the number of stars increases by 2.
- In the lower half, the number of spaces in each row increases by 1, while the number of stars decreases by 2.
- The `main` function prompts the user to enter a number and calls the `printDiamondPattern` function to print the diamond pattern.

Time Complexity: The time complexity of this code is O(n^2), where n is the number entered by the user. This is because there are two nested loops used to print the diamond pattern.

Space Complexity: The space complexity of this code is O(1), as there are no additional data structures used that depend on the input size.

Sample Input:
Enter a number for the widest point of the diamond: 5

Sample Output:
    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *

*/


package main

import (
	"fmt"
	"math"
)

func printSpaces(num int) {
	for i := 1; i <= num; i++ {
		fmt.Print(" ")
	}
}

func printStars(num int) {
	for i := 1; i <= num; i++ {
		fmt.Print("* ")
	}
}

func printDiamondPattern(n int) {
	// upper half of the diamond
	for i := 0; i < n; i++ {
		printSpaces(n - i - 1)
		printStars(2*i + 1)
		fmt.Println()
	}

	// lower half of the diamond
	for i := 0; i < n-1; i++ {
		printSpaces(i + 1)
		printStars(2*(n-i-1) - 1)
		fmt.Println()
	}
}

func main() {
	var num int
	fmt.Print("Enter a number for the widest point of the diamond: ")
	fmt.Scan(&num)

	printDiamondPattern(num)
}

